dali/operators/reader/fits_reader_op.cc

```diff
@@ -0,0 +1,152 @@
+
+DALI_REGISTER_OPERATOR(readers__Fits, FitsReaderCPU, CPU);
+
+DALI_SCHEMA(readers__Fits)
+    .DocStr(R"(Reads Fits image HDUs from a directory.
+
+This operator can be used in the following modes:
+
+1. Read all files from a directory indicated by ``file_root`` that match given ``file_filter``.
+2. Read file names from a text file indicated in ``file_list`` argument.
+3. Read files listed in ``files`` argument.
+
+.. note::
+  The ``gpu`` backend requires cuFile/GDS support (418.x driver family or newer). which is
+  shipped with the CUDA toolkit starting from CUDA 11.4. Please check the GDS documentation
+  for more details.
+
+  The ``gpu`` reader reads the files in chunks. The size of the chunk can be controlled
+  process-wide with an environment variable ``DALI_GDS_CHUNK_SIZE``. Valid values are powers of 2
+  between 4096 and 16M, with the default being 2M. For convenience, the value can be specified
+  with a k or M suffix, applying a multiplier of 1024 and 2^20, respectively.
+)")
+    .NumInput(0)
+    .NumOutput(1)  // (Arrays)
+    .AddOptionalArg<string>("file_root",
+                            R"(Path to a directory that contains the data files.
+
+If not using ``file_list`` or ``files``, this directory is traversed to discover the files.
+``file_root`` is required in this mode of operation.)",
+                            nullptr)
+    .AddOptionalArg(
+        "file_filter",
+        R"(If a value is specified, the string is interpreted as glob string to filter the
+list of files in the sub-directories of the ``file_root``.
+
+This argument is ignored when file paths are taken from ``file_list`` or ``files``.)",
+        "*.npy")
+    .AddOptionalArg<string>("file_list",
+                            R"(Path to a text file that contains filenames (one per line)
+where the filenames are relative to the location of that file or to ``file_root``, if specified.
+
+This argument is mutually exclusive with ``files``.)",
+                            nullptr)
+    .AddOptionalArg("shuffle_after_epoch",
+                    R"(If set to True, the reader shuffles the entire dataset after each epoch.
+
+``stick_to_shard`` and ``random_shuffle`` cannot be used when this argument is set to True.)",
+                    false)
+    .AddOptionalArg<vector<string>>("files", R"(A list of file paths to read the data from.
+
+If ``file_root`` is provided, the paths are treated as being relative to it.
+
+This argument is mutually exclusive with ``file_list``.)",
+                                    nullptr)
+    .AddOptionalArg("register_buffers",
+                    R"code(Applies **only** to the ``gpu`` backend type.
+
+.. warning::
+    This argument is temporarily disabled and left for backward compatibility.
+    It will be reenabled in the future releases.
+
+If true, the device I/O buffers will be registered with cuFile. It is not recommended if the sample
+sizes vary a lot.)code",
+                    true)
+    .AddParent("LoaderBase");
```

Comment:
As discussed, I would ask for following adjustments to the API:
* allow the user to specify which HDU indices they want to read. This would ensure easy mapping of 1 sample (potentially consisting of several images) to 1 file as well as allowing us to extend the reader with ability to load other types of HDUs (like some arrays with metadata). This can be something like `hdu_indices=[1, 4]` argument

Each requested HDU should map to one output. You can handle this by specifying a lambda via `AdditionalOutputsFn` that would calculate how many outputs are there.

Maybe we can have some kind of default right now, for example if the user doesn't provide any required outpus, they want the first image HDU so we always have at least one output.

* add the ability to specify the `dtype` and `ndim` of each output - this should be mandatory, it should accept one or more values (the same number as requested HDUs to return). We should check if the user provided correct values and raise a meaningful error if not. 

* Optionally, we can add some argument that allows the user to ensure if they are reading HDU with a correct extension.

Reply:
Okay, I have a few questions then: Given that reader can take as an argument path to a directory with potentially multiple files, should the `hdu_indices` by a vector<vector<int>> for each file?

Reply:
I think we should stick with the same vector of indices for all files that we are reading.